File: rust/src/keyring.rs
use anyhow::Result;
use std::fs;
use std::path::PathBuf;

use crate::display;
use crate::paths::get_profile_dir;

pub const SERVICE_NAME: &str = "arc";

fn get_fallback_path(profile_name: &str) -> PathBuf {
    get_profile_dir(profile_name).join(".api-key")
}

fn write_fallback_file(file_path: &PathBuf, secret: &str) -> Result<()> {
    if let Some(dir) = file_path.parent() {
        fs::create_dir_all(dir)?;
    }
    fs::write(file_path, secret)?;

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let perms = fs::Permissions::from_mode(0o600);
        fs::set_permissions(file_path, perms)?;
    }

    Ok(())
}

pub fn store_secret(profile_name: &str, secret: &str) -> Result<()> {
    let entry = keyring::Entry::new(SERVICE_NAME, profile_name);
    match entry {
        Ok(e) => match e.set_password(secret) {
            Ok(_) => return Ok(()),
            Err(err) => {
                display::warn(&format!(
                    "Failed to store secret in OS keyring ({}). Falling back to plaintext file.",
                    err
                ));
            }
        },
        Err(err) => {
            display::warn(&format!(
                "OS keyring not available ({}). Storing API key in plaintext file.",
                err
            ));
        }
    }

    let fallback = get_fallback_path(profile_name);
    write_fallback_file(&fallback, secret).unwrap_or_else(|e| {
        display::warn(&format!("Failed to store secret in fallback file: {}", e));
    });

    Ok(())
}

pub fn retrieve_secret(profile_name: &str) -> Option<String> {
    // Try OS keyring first
    if let Ok(entry) = keyring::Entry::new(SERVICE_NAME, profile_name) {
        if let Ok(password) = entry.get_password() {
            return Some(password);
        }
    }

    // Fallback: read from file
    let fallback = get_fallback_path(profile_name);
    if fallback.exists() {
        match fs::read_to_string(&fallback) {
            Ok(content) => return Some(content),
            Err(e) => {
                display::warn(&format!(
                    "Failed to read secret from fallback file: {}",
                    e
                ));
            }
        }
    }

    None
}

pub fn delete_secret(profile_name: &str) {
    // Try OS keyring
    if let Ok(entry) = keyring::Entry::new(SERVICE_NAME, profile_name) {
        let _ = entry.delete_credential();
    }

    // Also remove fallback file if present
    let fallback = get_fallback_path(profile_name);
    if fallback.exists() {
        if let Err(e) = fs::remove_file(&fallback) {
            display::warn(&format!(
                "Failed to delete secret fallback file: {}",
                e
            ));
        }
    }
}

File: packages/core/src/keyring.ts
import fs from "node:fs";
import path from "node:path";
import { getProfileDir } from "./paths.js";
import { writeLogEvent } from "./logging.js";

export const SERVICE_NAME = "arc";

interface KeyringEntry {
  setPassword(password: string): void;
  getPassword(): string;
  deletePassword(): void;
}

interface KeyringModule {
  Entry: new (service: string, account: string) => KeyringEntry;
}

let keyringModule: KeyringModule | null = null;
let keyringChecked = false;

async function loadKeyring(): Promise<KeyringModule | null> {
  if (keyringChecked) {
    return keyringModule;
  }
  keyringChecked = true;
  try {
    const mod = (await import("@napi-rs/keyring")) as KeyringModule;
    keyringModule = mod;
    return mod;
  } catch {
    keyringModule = null;
    return null;
  }
}

export function isKeyringAvailable(): boolean {
  return keyringModule !== null;
}

function getFallbackPath(profileName: string): string {
  return path.join(getProfileDir(profileName), ".api-key");
}

function writeFallbackFile(filePath: string, secret: string): void {
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, secret, { encoding: "utf-8", mode: 0o600 });
  if (process.platform !== "win32") {
    fs.chmodSync(filePath, 0o600);
  }
}

function logWarning(message: string): void {
  writeLogEvent({
    level: "warn",
    component: "keyring",
    action: "warning",
    message,
  });
}

export async function storeSecret(profileName: string, secret: string): Promise<void> {
  const kr = await loadKeyring();
  if (kr) {
    try {
      const entry = new kr.Entry(SERVICE_NAME, profileName);
      entry.setPassword(secret);
      return;
    } catch {
      logWarning("Failed to store secret in OS keyring. Falling back to plaintext file.");
    }
  } else {
    logWarning(
      "OS keyring not available (@napi-rs/keyring not installed). Storing API key in plaintext file."
    );
  }

  try {
    writeFallbackFile(getFallbackPath(profileName), secret);
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    logWarning(`Failed to store secret in fallback file: ${message}`);
  }
}

export async function retrieveSecret(profileName: string): Promise<string | null> {
  const kr = await loadKeyring();
  if (kr) {
    try {
      const entry = new kr.Entry(SERVICE_NAME, profileName);
      return entry.getPassword();
    } catch {
      // fall through
    }
  }

  try {
    const fallbackPath = getFallbackPath(profileName);
    if (fs.existsSync(fallbackPath)) {
      return fs.readFileSync(fallbackPath, "utf-8");
    }
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    logWarning(`Failed to read secret from fallback file: ${message}`);
  }

  return null;
}

export async function deleteSecret(profileName: string): Promise<void> {
  const kr = await loadKeyring();
  if (kr) {
    try {
      const entry = new kr.Entry(SERVICE_NAME, profileName);
      entry.deletePassword();
    } catch {
      // ignore
    }
  }

  try {
    const fallbackPath = getFallbackPath(profileName);
    if (fs.existsSync(fallbackPath)) {
      fs.unlinkSync(fallbackPath);
    }
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    logWarning(`Failed to delete secret fallback file: ${message}`);
  }
}
